### PR TITLE
Polkadot v0.9.38

### PR DESCRIPTION
**Pull request by wangjj9219**

### CODE REVIEW

File: runtime/acala/src/xcm_config.rs [hunk 1]
```diff
@@ -281,16 +306,16 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
...
 		}
 
 		match location {
 			MultiLocation {
 				parents,
-				interior: X2(Parachain(para_id), GeneralKey(key)),
+				interior: X2(Parachain(para_id), GeneralKey { data, length }),
 			} if parents == 1 => {
-				match (para_id, &key.into_inner()[..]) {
+				match (para_id, &data[..length as usize]) {
```
Comment (zjb0807) [inline runtime/acala/src/xcm_config.rs:318]: same here

File: runtime/acala/src/xcm_config.rs [hunk 2]
```diff
@@ -313,9 +338,9 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
 			// adapt for re-anchor canonical location: https://github.com/paritytech/polkadot/pull/4470
 			MultiLocation {
 				parents: 0,
-				interior: X1(GeneralKey(key)),
+				interior: X1(GeneralKey { data, length }),
 			} => {
-				let key = &key.into_inner()[..];
+				let key = &data[..length as usize];
```
Comment (zjb0807) [inline runtime/acala/src/xcm_config.rs:343]: same here

File: runtime/karura/src/lib.rs [hunk 3]
```diff
@@ -1803,8 +1801,11 @@ pub type Executive = frame_executive::Executive<
 	Runtime,
 	AllPalletsWithSystem,
 	(
-		pallet_balances::migration::MigrateToTrackInactive<Runtime, xcm_config::CheckingAccount>,
-		pallet_scheduler::migration::v4::CleanupAgendas<Runtime>,
+		// "Use 2D weights in XCM v3" <https://github.com/paritytech/polkadot/pull/6134>
+		pallet_xcm::migration::v1::MigrateToV1<Runtime>,
+		// Note: The following Migrations do not use the StorageVersion feature, must to be removed after the upgrade
+		module_asset_registry::migrations::MigrateV1MultiLocationToV3<Runtime>,
+		module_xcm_interface::migrations::MigrateXcmDestWeightAndFee<Runtime>,
```
Comment (zjb0807) [inline runtime/karura/src/lib.rs:1808]: same here

File: runtime/karura/src/xcm_config.rs [hunk 4]
```diff
@@ -386,16 +409,16 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
...
 		}
 
 		match location {
 			MultiLocation {
 				parents: 1,
-				interior: X2(Parachain(para_id), GeneralKey(key)),
+				interior: X2(Parachain(para_id), GeneralKey { data, length }),
 			} => {
-				match (para_id, &key.into_inner()[..]) {
+				match (para_id, &data[..length as usize]) {
```
Comment (zjb0807) [inline runtime/karura/src/xcm_config.rs:421]: same here

File: runtime/karura/src/xcm_config.rs [hunk 5]
```diff
@@ -426,9 +449,10 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
 			// adapt for re-anchor canonical location: https://github.com/paritytech/polkadot/pull/4470
 			MultiLocation {
 				parents: 0,
-				interior: X1(GeneralKey(key)),
+				interior: X1(GeneralKey { data, length }),
 			} => {
-				let currency_id = CurrencyId::decode(&mut &*key.into_inner()).ok()?;
+				let key = &data[..length as usize];
```
Comment (zjb0807) [inline runtime/karura/src/xcm_config.rs:454]: same here

File: runtime/mandala/src/xcm_config.rs [hunk 6]
```diff
@@ -289,9 +317,9 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
 			// adapt for re-anchor canonical location: https://github.com/paritytech/polkadot/pull/4470
 			MultiLocation {
 				parents: 0,
-				interior: X1(GeneralKey(key)),
+				interior: X1(GeneralKey { data, length }),
 			} => {
-				let key = &key.into_inner()[..];
+				let key = &data[..length as usize];
```
Comment (zjb0807) [inline runtime/mandala/src/xcm_config.rs:322]: same here

File: runtime/mandala/src/xcm_config.rs [hunk 7]
```diff
@@ -263,17 +291,17 @@ impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
...
 
 		match location {
 			MultiLocation {
 				parents,
-				interior: X2(Parachain(para_id), GeneralKey(key)),
+				interior: X2(Parachain(para_id), GeneralKey { data, length }),
 			} if parents == 1 && ParaId::from(para_id) == ParachainInfo::get() => {
 				// decode the general key
-				let key = &key.into_inner()[..];
+				let key = &data[..length as usize];
```
Comment (zjb0807) [inline runtime/mandala/src/xcm_config.rs:304]: same here

File: runtime/mandala/src/lib.rs [hunk 8]
```diff
@@ -1904,8 +1902,11 @@ pub type Executive = frame_executive::Executive<
 	Runtime,
 	AllPalletsWithSystem,
 	(
-		pallet_balances::migration::MigrateToTrackInactive<Runtime, xcm_config::CheckingAccount>,
-		pallet_scheduler::migration::v4::CleanupAgendas<Runtime>,
+		// "Use 2D weights in XCM v3" <https://github.com/paritytech/polkadot/pull/6134>
+		pallet_xcm::migration::v1::MigrateToV1<Runtime>,
+		// Note: The following Migrations do not use the StorageVersion feature, must to be removed after the upgrade
+		module_asset_registry::migrations::MigrateV1MultiLocationToV3<Runtime>,
+		module_xcm_interface::migrations::MigrateXcmDestWeightAndFee<Runtime>,
```
Comment (zjb0807) [inline runtime/mandala/src/lib.rs:1909]: same here

File: runtime/acala/src/xcm_config.rs [hunk 9]
```diff
@@ -179,8 +181,17 @@ impl xcm_executor::Config for XcmConfig {
 		NativeTokenExistentialDeposit,
 		ExistentialDeposits,
 	>;
+	type AssetLocker = ();
+	type AssetExchanger = ();
 	type AssetClaims = PolkadotXcm;
 	type SubscriptionService = PolkadotXcm;
+	type PalletInstancesInfo = AllPalletsWithSystem;
+	type MaxAssetsIntoHolding = ConstU32<64>;
+	type FeeManager = ();
```
Comment (xlc) [inline runtime/acala/src/xcm_config.rs:190]: I think we need to implement this to deposit the fee to treasury

File: runtime/acala/src/xcm_config.rs [hunk 10]
```diff
@@ -212,11 +228,19 @@ impl pallet_xcm::Config for Runtime {
...
 	type RuntimeOrigin = RuntimeOrigin;
 	type RuntimeCall = RuntimeCall;
 	const VERSION_DISCOVERY_QUEUE_SIZE: u32 = 100;
 	type AdvertisedXcmVersion = pallet_xcm::CurrentXcmVersion;
+	type Currency = Balances;
+	type CurrencyMatcher = ();
+	type TrustedLockers = ();
+	type SovereignAccountOf = ();
+	type MaxLockers = ConstU32<8>;
+	type WeightInfo = pallet_xcm::TestWeightInfo; // TODO: config after polkadot impl WeightInfo for ()
```
Comment (xlc) [inline runtime/acala/src/xcm_config.rs:241]: we need to setup benchmarking for this

we can for now copy the weights from https://github.com/paritytech/cumulus/blob/66887505e6c5ef89ce9f88fc4ec368eb858b1b92/parachains/runtimes/assets/statemint/src/weights/pallet_xcm.rs#L92

File: runtime/common/src/xcm_impl.rs [hunk 11]
```diff
@@ -116,9 +108,9 @@ where
 			}
 		}
 		if !asset_traps.is_empty() {
-			X::drop_assets(origin, asset_traps.into());
+			X::drop_assets(origin, asset_traps.into(), context);
 		}
-		0
+		XcmWeight::from_ref_time(0)
```
Comment (xlc) [inline runtime/common/src/xcm_impl.rs:113]: need a TODO & issue

File: runtime/acala/src/xcm_config.rs (hunk 9 above)
Comment (wangjj9219) [reply runtime/acala/src/xcm_config.rs:190]: Currently Statemint/Statemine hasn't charged this fee, we will follow their implementation

**Review by xlc — APPROVED** [review]

**Pull request merged**